cam6_3_026: Support for the CAM Perturbed Parameter Ensemble (PPE)

Katetc opened the pull request:
Code to bring many hardcoded parameters to namelist options so that we can experiment with model sensitivity to them. The tag including these code changes will be used in the CAM PPE. 

Note: There could be updates to the CLM tag (to conform to the CLM PPE) and to the PUMAS tag (only to improve comment documentation) before this PR goes in. These changes will likely be done Wednesday 7-7.

fixes #313

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -3529,6 +3665,21 @@ clubb_up2_vp2_factor increases the values of up2 and vp2 at the surface.
 
+<entry id="clubb_detliq_rad" type="real" category="pblrad"
+       group="clubb_params_nl" valid_values="" >
+Radius of detrained liquid drops as they are used in the CLUBB parameterization.
```
Units?

adamrher replied:
m

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -3529,6 +3665,21 @@ clubb_up2_vp2_factor increases the values of up2 and vp2 at the surface.
+
+<entry id="clubb_detice_rad" type="real" category="pblrad"
+       group="clubb_params_nl" valid_values="" >
+Radius of detrained ice drops as they are used in the CLUBB parameterization.
```
Units?

adamrher replied:
m

gold2718 commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -3529,6 +3665,21 @@ clubb_up2_vp2_factor increases the values of up2 and vp2 at the surface.
+
+<entry id="clubb_detphase_lowtemp" type="real" category="pblrad"
+       group="clubb_params_nl" valid_values="" >
+Temperature at which detrained water is classified as entirely ice (no liquid) 
+in the CLUBB parameterization.
```
Units?

adamrher replied:
K

gold2718 reviewed (changes requested):
The new namelist items really need some documentation (see notes). Otherwise, this looks okay.

andrewgettelman commented:
Go ahead without me. Can’t see it for a week. 

Hopefully this also has Trude’s CLUBB additions.

Regards,

Andrew

Sent from my iPhone (sorry for typos)
Andrew Gettelman
National Center for Atmospheric Research
1850 Table Mesa Dr., Boulder, CO 80304
***@***.***
+1-303-497-1887

> On Jul 6, 2021, at 6:37 PM, Kate Thayer-Calder ***@***.***> wrote:
> 
> ﻿
> @Katetc requested your review on: #398 Support for the CAM Perturbed Parameter Ensemble (PPE) .
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.

andrewgettelman commented:
We do NOT need to document the sensitivities. We don’t know what they are: that’s why we are doing this.

Fine to document more specifically if you want.

Else check it in and I can add detail next week.

Thanks,

Andrew

Sent from my iPhone (sorry for typos)
Andrew Gettelman
National Center for Atmospheric Research
1850 Table Mesa Dr., Boulder, CO 80304
***@***.***
+1-303-497-1887

> On Jul 6, 2021, at 8:50 PM, goldy ***@***.***> wrote:
> 
> ﻿
> @gold2718 requested changes on this pull request.
> 
> The new namelist items really need some documentation (see notes). Otherwise, this looks okay.
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Maximum allowed ice number concentration
> +Default: 1000000.e3
> This is a weird way to write 10^9. Why not 1.0e9?
> Also, what are the units here?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Accretion enhancment factor
> +Default: 1
> This needs more documentation. What is enhanced if you change this value? By how much?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Autoconversion prefactor
> +Default: 0.001
> This needs more documentation. What changes if you change this value? By how much? What are the units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Autoconversion nd exponent
> +Default: -1.1
> This needs more documentation. What happens if you change this value? By how much?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Autoconversion lwp exponent
> +Default: 2.47
> This needs more documentation. What happens if you change this value?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Size of droplets feezing homogeneously 
> +Default: 1
> This needs more documentation. What are the units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Ice fall speed factor
> +Default: 1
> This needs more documentation. What happens if you change this value? By how much? Any units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Ice effective radius factor
> +Default: 25.e-6
> This needs more documentation. What happens if you change this value? By how much? Any units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +ccn activated number scale factor
> +Default: 1
> This needs more documentation. What happens if you change this value? By how much? Any units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Scaling factor for subgrid vertical velocity (liquid)
> +Default: 1
> This needs more documentation. What happens if you change this value? By how much? Any units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Scaling factor for subgrid vertical velocity (ice)
> +Default: 1
> This needs more documentation. What happens if you change this value? By how much? Any units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Minimum subgrid vertical velocity (liquid) ms-1
> +Default: 0.2
> This needs more documentation. What happens if you change this value? By how much? Any units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Minimum subgrid vertical velocity (ice) ms-1
> +Default: 0.001
> This needs more documentation. What happens if you change this value? By how much? Any units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Tunable entrainment rate in ZM deep convection scheme.                     
> +Default: set by build-namelist
> Any units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Tunable parcel temperature perturbation  in ZM deep convection scheme.
> +Default: set by build-namelist
> Units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Tunable triggering threshold for convection in ZM deep scheme.
> +Default: set by build-namelist
> Units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > @@ -3529,6 +3665,21 @@ clubb_up2_vp2_factor increases the values of up2 and vp2 at the surface.
>  Coefficient for gustiness near ground.
>  </entry>
>  
> +<entry id="clubb_detliq_rad" type="real" category="pblrad"
> +       group="clubb_params_nl" valid_values="" >
> +Radius of detrained liquid drops as they are used in the CLUBB parameterization.
> Units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > @@ -3529,6 +3665,21 @@ clubb_up2_vp2_factor increases the values of up2 and vp2 at the surface.
>  Coefficient for gustiness near ground.
>  </entry>
>  
> +<entry id="clubb_detliq_rad" type="real" category="pblrad"
> +       group="clubb_params_nl" valid_values="" >
> +Radius of detrained liquid drops as they are used in the CLUBB parameterization.
> +</entry>
> +
> +<entry id="clubb_detice_rad" type="real" category="pblrad"
> +       group="clubb_params_nl" valid_values="" >
> +Radius of detrained ice drops as they are used in the CLUBB parameterization.
> Units?
> 
> In bld/namelist_files/namelist_definition.xml:
> 
> > +Temperature at which detrained water is classified as entirely ice (no liquid) 
> +in the CLUBB parameterization.
> Units?
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or unsubscribe.

adamrher commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2945,6 +3029,23 @@ Default: =&gt; 1 for CAM6;
 
+<entry id="zmconv_dmpdz" type="real" category="conv"
+       group="zmconv_nl" valid_values="" >
+Tunable entrainment rate in ZM deep convection scheme.                     
+Default: set by build-namelist
```
peanut gallery here. It's 1/m

adamrher commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2945,6 +3029,23 @@ Default: =&gt; 1 for CAM6;
+
+<entry id="zmconv_tiedke_add" type="real" category="conv"
+       group="zmconv_nl" valid_values="" >
+Tunable parcel temperature perturbation  in ZM deep convection scheme.
+Default: set by build-namelist
```
temperature, so K

adamrher commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2945,6 +3029,23 @@ Default: =&gt; 1 for CAM6;
+
+<entry id="zmconv_capelmt" type="real" category="conv"
+       group="zmconv_nl" valid_values="" >
+Tunable triggering threshold for convection in ZM deep scheme.
+Default: set by build-namelist
```
J/kg

cacraigucar commented on cime_config/usermods_dirs/scam_twp06/user_nl_cam:
```diff
@@ -12,3 +12,5 @@ scm_relax_top_p        = 200.
 scm_relax_tau_bot_sec                = 864000.
 scm_relax_tau_top_sec                = 172800.
+iradlw = 1
+iradsw = 1
```
The comparison says there is no "newline at the end of the file".  I've never seen this before and just suggest you check the last line to make sure there isn't something funky at the end.

cacraigucar commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -743,6 +782,11 @@ subroutine clubb_readnl(nlfile)
+  if(clubb_detliq_rad == unset_r8) call endrun(sub//": FATAL: clubb_detliq_rad not set")
+  if(clubb_detice_rad == unset_r8) call endrun(sub//": FATAL: clubb_detice_rad not set")
+  if(clubb_detphase_lowtemp == unset_r8) call endrun(sub//": FATAL: clubb_detphase_lowtemp not set")
+  if(clubb_detphase_lowtemp >= 268.15_r8) & 
+     call endrun(sub//": ERROR: clubb_detphase_lowtemp must be less than 268.15 K")
```
Should the error message say "less than or equal" or does the if statement need to revised?

gold2718 replied:
I think the `if` statement is correct. If they are equal, I think you might get a divide by zero issue down on line 3130.

@Katetc, what happens if `clubb_detphase_lowtemp` is very close to `265.15` (e.g., 265.1499999999)?

Katetc replied:
Later on, the equation is if the state temp is between 268.15 and low, then ( 268.15 - temp ) / (268.15 - low), which gives a fraction of the cloud water to be ice and the remainder is liquid. If the "low" value is very close to 268.15, and then the state temperature somehow manages to be between 268.15 and 268.1499, then the numerator and denominator would be equivalently sized. This math shouldn't blow up.

cacraigucar commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -3077,33 +3124,33 @@ subroutine clubb_tend_cam( &
             dum1 = 1.0_r8
          else
-            dum1 = ( 268.15_r8 - state1%t(i,k) ) / 30._r8 
+            dum1 = ( 268.15_r8 - state1%t(i,k) ) / ( 268.15_r8 - dt_low ) 
```
I keep seeing `268.15_r8` throughout the code.  While you didn't introduce it, you did utilize it a couple of times.  I would suggest it would make sense to create a parameter with this value and use it throughout.

adamrher commented:
Hopefully I'm being helpful --I added some the units for most of what steve flagged. Also, I sound like a broken record, but the namelists ```cldfrc_dp1``` and ```cldfrc_dp2``` are hard coded into clubb_intr, since those namelists are linked up with macropdriver (which we don't use in cam6 physics). So a change like this needs to be added to clubb_intr:

```
         (-) deepcu(i,k) = max(0.0_r8,min(0.1_r8*log(1.0_r8+500.0_r8*(cmfmc(i,k+1)-cmfmc_sh(i,k+1))),0.6_r8))
         (+) deepcu(i,k) = max(0.0_r8,min(dp1*log(1.0_r8+dp2*(cmfmc(i,k+1)-cmfmc_sh(i,k+1))),0.6_r8))
```

with ```dp1``` and ```dp2``` linked up to those namelists. I have made these changes in my pending PR: https://github.com/ESCOMP/CAM/pull/352

cacraigucar commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -1794,6 +1794,12 @@
 <clubb_C2thl                                      > 1.0     </clubb_C2thl>
 <clubb_C2rtthl                                    > 1.3     </clubb_C2rtthl>
 <clubb_C4                                         > 5.2     </clubb_C4>
+<clubb_c6rt                                       > 4.0     </clubb_c6rt>
+<clubb_c6rtb                                      > 6.0     </clubb_c6rtb>
+<clubb_c6rtc                                      > 1.0     </clubb_c6rtc>
```
Are all of these newly introduced default values the default that CLUBB is currently using (i.e. will this change answers for any existing CAM6 runs)?  I don't think this tag should be changing values for current code.

Katetc replied:
These default values are the same ones currently used by clubb as of cam6_3_025. This tag does change answers slightly because of one microphysics parameter that went from a hard-coded volume to a volume calculated based on radius. It is non-climate changing.

cacraigucar commented on src/physics/cam/micro_mg_cam.F90:
```diff
@@ -122,6 +122,19 @@ module micro_mg_cam
 character(len=16) :: micro_mg_precip_frac_method = 'max_overlap' ! type of precipitation fraction method
 
 real(r8)          :: micro_mg_berg_eff_factor    = 1.0_r8        ! berg efficiency factor
+!++ trude for namelist
```
Remove this comment from here and all places below

cacraigucar commented on src/physics/cam/micro_mg_cam.F90:
```diff
@@ -122,6 +122,19 @@ module micro_mg_cam
+real(r8)          :: micro_mg_effi_factor = 1.0_r8  ! ice effective radius factor
+real(r8)          :: micro_mg_iaccr_factor = 1.0_r8 ! ice accretion of cloud droplet
+real(r8)          :: micro_mg_max_nicons = 100000.e3_r8 ! max allowed ice number concentration
+!-- trude
```
Remove this comment here and all similar places below

cacraigucar commented on src/physics/cam/micro_mg_cam.F90:
```diff
@@ -122,6 +122,19 @@ module micro_mg_cam
 
 real(r8)          :: micro_mg_berg_eff_factor    = 1.0_r8        ! berg efficiency factor
+!++ trude for namelist
+real(r8)          :: micro_mg_accre_enhan_fact    = 1.0_r8        ! accretion enhancment factor
+real(r8)           :: micro_mg_autocon_fact = 0.01_r8 ! autoconversion prefactor
+real(r8)           :: micro_mg_autocon_nd_exp = -1.1_r8 ! autoconversion nd exponent
+real(r8)           :: micro_mg_autocon_lwp_exp = 2.47_r8    ! autoconversion lwp exponent      
+real(r8)           :: micro_mg_homog_size = 25.e-6_r8! size of freezing homogeneous ice
+
+real(r8)          :: micro_mg_vtrmi_factor = 1.0_r8  ! ice fall speed factor
+real(r8)          :: micro_mg_effi_factor = 1.0_r8  ! ice effective radius factor
+real(r8)          :: micro_mg_iaccr_factor = 1.0_r8 ! ice accretion of cloud droplet
+real(r8)          :: micro_mg_max_nicons = 100000.e3_r8 ! max allowed ice number concentration
```
These should probably all be set to UNSET here and depend on the defaults being set in namelist_defaults.

cacraigucar commented on src/physics/cam/micro_mg_cam.F90:
```diff
@@ -897,6 +955,10 @@ subroutine micro_mg_cam_init(pbuf2d)
            micro_mg_do_hail,micro_mg_do_graupel, &
            microp_uniform, do_cldice, use_hetfrz_classnuc, &
            micro_mg_precip_frac_method, micro_mg_berg_eff_factor, &
+           micro_mg_accre_enhan_fact , &  !++ trude
```
Remove the !++ trude comments here and below

cacraigucar commented on src/physics/cam/micro_mg_cam.F90:
```diff
@@ -2575,7 +2637,7 @@ subroutine micro_mg_cam_tend_pack(state, ptend, dtime, pbuf, mgncol, mgcols, nle
               packed_mnuccc,  packed_mnucct,  packed_msacwi,  &
               packed_psacws,  packed_bergs,   packed_berg,    &
-              packed_melt,    packed_meltstot, packed_meltgtot,           packed_homo,            &
+              packed_melt, packed_meltstot, packed_meltgtot,            packed_homo,            &
```
This line lost some spaces and no longer lines up like it once did.

cacraigucar commented on src/physics/cam/micro_mg_cam.F90:
```diff
@@ -2601,7 +2663,7 @@ subroutine micro_mg_cam_tend_pack(state, ptend, dtime, pbuf, mgncol, mgcols, nle
       end select
 
-      call handle_errmsg(errstring, subname="micro_mg_tend")
+      call handle_errmsg(errstring, subname="micro_mg_  tend")
```
Spaces before `tend` need to be removed

cacraigucar commented on src/physics/cam/microp_aero.F90:
```diff
@@ -316,13 +322,20 @@ subroutine microp_aero_readnl(nlfile)
    ! Namelist variables
-   real(r8) :: microp_aero_bulk_scale = 2._r8  ! prescribed aerosol bulk sulfur scale factor
- 
+   real(r8) :: microp_aero_bulk_scale = 2._r8 ! prescribed aerosol bulk sulfur scale factor
+   real(r8) :: microp_aero_npccn_scale = 1._r8  ! prescribed aerosol bulk sulfur scale factor
+   real(r8) :: microp_aero_wsub_scale = 1._r8  ! subgrid vertical velocity (liquid) scale factor
+   real(r8) :: microp_aero_wsubi_scale = 1._r8  ! subgrid vertical velocity (ice) scale factor
+   real(r8) :: microp_aero_wsub_min = 0.2_r8  ! subgrid vertical velocity (liquid) minimum
+   real(r8) :: microp_aero_wsubi_min = 0.001_r8  ! subgrid vertical velocity (ice) minimum
```
These should be set to UNSET and use the namelist_default values

cacraigucar commented on src/physics/cam/microp_aero.F90:
```diff
@@ -342,10 +355,20 @@ subroutine microp_aero_readnl(nlfile)
 #ifdef SPMD
    ! Broadcast namelist variable
    call mpibcast(microp_aero_bulk_scale, 1, mpir8, 0, mpicom)
+   call mpibcast(microp_aero_npccn_scale, 1, mpir8, 0, mpicom)
+   call mpibcast(microp_aero_wsub_scale, 1, mpir8, 0, mpicom)
+   call mpibcast(microp_aero_wsubi_scale, 1, mpir8, 0, mpicom)
+   call mpibcast(microp_aero_wsub_min, 1, mpir8, 0, mpicom)
+   call mpibcast(microp_aero_wsubi_min, 1, mpir8, 0, mpicom)
```
We've been updating to the newer mpi_bcast calls as we encounter the old ones.  Please update them and add error handling.

cacraigucar commented on src/physics/cam/microp_aero.F90:
```diff
@@ -523,8 +546,8 @@ subroutine microp_aero_run ( &
    ! Set minimum values above top_lev.
-   wsub(:ncol,:top_lev-1)  = 0.20_r8
-   wsubi(:ncol,:top_lev-1) = 0.001_r8
+   wsub(:ncol,:top_lev-1)  = wsub_min  !0.20_r8
+   wsubi(:ncol,:top_lev-1) = wsubi_min !0.001_r8
```
The comments should be removed as these are namelist values and are subject to changed values.

gold2718 commented:
> We do NOT need to document the sensitivities. We don’t know what they are: that’s why we are doing this.
> Fine to document more specifically if you want.

Right. I do not expect that we can document values (yet) but we should document what changing these values changes in the computation.

cacraigucar commented on src/physics/cam/zm_conv.F90:
```diff
@@ -40,7 +40,8 @@ module zm_conv
    real(r8) rl         ! wg latent heat of vaporization.
    real(r8) cpres      ! specific heat at constant pressure in j/kg-degk.
-   real(r8), parameter :: capelmt = 70._r8  ! threshold value for cape for deep convection.
+   real(r8) :: capelmt = 70._r8  ! namelist configurable: 
```
Remove the initialization value as it is set by the namelist or initlialize to UNSET

cacraigucar commented on src/physics/cam/zm_conv.F90:
```diff
@@ -72,14 +73,16 @@ module zm_conv
    integer  limcnv       ! top interface level limit for convection
 
-   real(r8),parameter ::  tiedke_add = 0.5_r8   
+   real(r8) :: tiedke_add = 0.5_r8     ! New namelist configurable
+   real(r8) :: dmpdz_param = -1.e-3_r8 ! New namelist configurable
```
Remove the initializations or initialize to UNSET for these two variables as well.

cacraigucar commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2539,6 +2539,60 @@ Efficiency factor for berg
 
+<entry id="micro_mg_accre_enhan_fact" type="real" category="microphys"
+       group="micro_mg_nl" valid_values="" >
+Accretion enhancment factor
+Default: 1
```
@Katetc These requests are being made as this file autogenerates the namelist web page that users refer to when setting namelist values.  Because of this, this information is important to have accurate and complete.

Katetc commented:
Hi all. I made a first pass at addressing these comments, but I know there are still some more to do (especially micro_mg_cam). @andrewgettelman, yes, this does include Trude's additions of c6thl and c6thlb in clubb. @adamrher Thank you! Those units were super helpful. And I totally stole your code for dp1 and dp2 wholesale from your PR. Please have a look and make sure I've got it in there correctly, but I think that is addressed now.

cacraigucar commented on src/physics/cam/microp_aero.F90:
```diff
@@ -615,6 +652,15 @@ subroutine microp_aero_run ( &
 
       npccn(:ncol,:) = nctend_mixnuc(:ncol,:)
 
+      do k = 1, pver
+         do i = 1, ncol
+            if(not(is_nan(npccn(i,k)))) then
+               npccn(i,k) = npccn(i,k) * npccn_scale
+            end if
+         enddo
+      enddo
+
+
```
If Francis' proposed initlialization of tendnd in dropmixnuc resolves the NaN problem, please remove this section of code.

gold2718 commented on src/physics/cam/clubb_intr.F90:
```diff
@@ -3150,7 +3154,7 @@ subroutine clubb_tend_cam( &
                                      / (4._r8*3.14_r8*di_rad**3*500._r8) + & ! Deep    Convection
                                      3._r8 * (                         dlf2(i,k)    *  dum1 ) &
-                                     / (4._r8*3.14_r8*50.e-6**3*500._r8)     ! Shallow Convection
+                                     / (4._r8*3.14_r8*50.e-6_r8**3*500._r8)     ! Shallow Convection
```
Is this version of `500._r8` also `dp2`?
Is `50.e-6_r8` something that should have a name?
Why is `3.14_r8` being used here instead of `pi`? Is it not supposed to be `pi`?

Katetc replied:
I just added the _r8 here so that the check test would pass on Izumi. This code is in no way related to the changes for the PPE. I have no good answers to these questions, though I suspect they might be "no, probably, ask Pete, and it is probably Pi, but not scoped to rewrite clubb_inter in this PR."

andrewgettelman replied:
Agreed: we need to get this done, not fix clubb_intr issues. Can be done later in a different issue or PR.

gold2718 replied:
Agreed (which is why it is a question and not a request).
This is a good item to address during conversion to the CCPP.

gold2718 reviewed (changes requested):
I have some questions that *could* result in code changes but they are not required.
I _would_ like us to be using standard unit language so that we can use them with the CCPP. Since these changes are in XML comment blocks, it should not require re-running all the tests (just run one test or run xmllint).

nusbaume reviewed (changes requested):
Looks good to me, although I do have some suggestions which will hopefully be easy to implement.

Katetc commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -1931,8 +1941,49 @@
-<micro_mg_berg_eff_factor                  >  1.0D0    </micro_mg_berg_eff_factor>
-<micro_mg_berg_eff_factor microphys="mg2"  >  1.0D0    </micro_mg_berg_eff_factor>
+<micro_mg_berg_eff_factor                  >  1.D0    </micro_mg_berg_eff_factor>
+<micro_mg_berg_eff_factor microphys="mg2"  >  1.D0    </micro_mg_berg_eff_factor>
```
Yes, we would like the ability to change these to tune MG2 and MG3 separately.

Katetc commented on bld/namelist_files/namelist_defaults_cam.xml:
```diff
@@ -1931,8 +1941,49 @@
+<micro_mg_iaccr_factor                  >  1.D0    </micro_mg_iaccr_factor>
+<micro_mg_iaccr_factor microphys="mg2"  >  1.D0    </micro_mg_iaccr_factor>
+
+<micro_mg_max_nicons                  >  100000.D3    </micro_mg_max_nicons>
```
Counting zeros is hard! :) It's actually D8. I'll fix it.

Katetc commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -3414,6 +3515,36 @@ C2 coef. for the thlp2_dp1 term
 C2 coef. for the rtpthlp_dp1 term
 </entry>
 
+<entry id="clubb_c6rt" type="real" category="pblrad"
```
Yes, these get twiddled-with quite a lot. Adding in all of the defaults here would take a lot more time than I'm supposed to be using for this PR.

cacraigucar replied:
I've always felt uncomfortable with labeling the defaults here and in the namelist_defaults file as it is too easy to get out-of-date and give the user misleading information.  I concur with Kate's decision to not add them, especially since they are in flux.

Katetc commented on src/physics/cam/zm_conv_intr.F90:
```diff
@@ -352,7 +363,8 @@ subroutine zm_conv_init(pref_edge)
     call zm_convi(limcnv,zmconv_c0_lnd, zmconv_c0_ocn, zmconv_ke, zmconv_ke_lnd, &
                   zmconv_momcu, zmconv_momcd, zmconv_num_cin, zmconv_org, &
-                  zmconv_microp, no_deep_pbl_in = no_deep_pbl)
+                  zmconv_microp, no_deep_pbl, zmconv_tiedke_add, &
```
I removed the optional logic so it will always be passed in now.

nusbaume commented on bld/namelist_files/namelist_definition.xml:
```diff
@@ -2539,6 +2539,60 @@ Efficiency factor for berg
+<entry id="micro_mg_max_nicons" type="real" category="microphys"
+       group="micro_mg_nl" valid_values="" >
+Maximum allowed ice number concentration
+Default: 1.0e9
```
I guess as an example of "we shouldn't have defaults listed in two different places", should this say `1.0e8` instead?

gold2718 replied:
This is inconsistent with what is in namelist_defaults.xml
Which is correct?

nusbaume reviewed (approved):
Thanks for answering my questions/concerns!  It looks great now, although I do have one last question.

gold2718 commented on doc/ChangeLog:
```diff
@@ -1,3 +1,108 @@
+Github PR URL: https://github.com/ESCOMP/CAM/pull/398
+
+Purpose of changes (include the issue number and title text for each relevant GitHub issue):
+- Adding in namelist parameters and supporting code to be used in the CAM PPE
```
Just needs issue number added:
```suggestion
- #313: Adding in namelist parameters and supporting code to be used in the CAM PPE
```

gold2718 commented on doc/ChangeLog:
```diff
@@ -1,3 +1,108 @@
+then copy the lines from the td.*.status files for the failed tests to the
+appropriate machine below.  All failed tests must be justified.
+
+cheyenne/intel/aux_cam:
+
+izumi/nag/aux_cam:
+
+izumi/pgi/aux_cam:
```
Please fill out this report (including the expected FAIL for the DAE test).

gold2718 reviewed (approved):
Tentatively approving. Please make the requested very minor changes before merging (and complete tests of course).
